metadata/metadata.go

```diff
@@ -91,15 +92,24 @@ func Pairs(kv ...string) MD {
 }
 
 // String implements the Stringer interface for pretty-printing a MD.
-// Ordering of the values is non-deterministic as it ranges over a map.
+// The value returned by this method is recommended only for diagnostic
+// purposes, and users are strongly discouraged from using this for anything
+// mission critical.
 func (md MD) String() string {
 	var sb strings.Builder
 	fmt.Fprintf(&sb, "MD{")
-	for k, v := range md {
-		if sb.Len() > 3 {
+	keys := make([]string, 0, len(md))
```

Comment:
Looks like the `fmt` package already sorts maps based on keys. See https://pkg.go.dev/fmt.

```
	// Maps formatted with %v show keys and values in their default formats.
	// The %#v form (the # is called a "flag" in this context) shows the map in
	// the Go source format. Maps are printed in a consistent order, sorted
	// by the values of the keys.
```

So, maybe all we need here is `fmt.Sprintf("%v", md)` or `fmt.Sprintf("%+v", md)`. Is it possible for you to check if that actually helps fix the problem you are seeing?

Reply:
I'm sure that will fix it, since that's what they were calling that led to `String()` being called now.  :)

But that presumably undoes the reason the `String()` method was added in the first place.  (Why did we add it BTW?  To make it look nicer?)

Reply:
Actually it looks like the main motivation of the PR that added it was to make `context.String()` show it, since the `valueContext`'s `String()` method only prints values that are strings or implement `Stringer` - otherwise it just prints something like `<not stringer>`.  So I think just `return fmt.Sprintf("%v", md)` would be fine here.

Reply:
and sadly it seems the underlying implementation uses a "fmtsort" Sort that's different from what we could expect and it lives in `internal`:
https://github.com/golang/go/blob/5a18e79687dea15680ff5f799b549fa0efd0cad9/src/internal/fmtsort/sort.go#L35-L49

Reply:
An easy solution is to assign to a local map in the `String()` @dims:
https://go.dev/play/p/dPXpqKW9Gfm

Reply:
I am happy to close this PR if you all want to do it some other way :)

Reply:
It's even easier:

```go
	return fmt.Sprint(map[string][]string(md))
```

Reply:
Done!